packages/backend-function/src/factory.test.ts

```diff
@@ -440,6 +440,36 @@ void describe('AmplifyFunctionFactory', () => {
     });
   });
 
+  void describe('logging options', () => {
+    void it('sets logging options', () => {
+      const lambda = defineFunction({
+        entry: './test-assets/default-lambda/handler.ts',
+        bundling: {
+          minify: false,
+        },
+        logging: {
+          format: 'json',
+          level: 'warn',
+          retention: '13 months',
+        },
+      }).getInstance(getInstanceProps);
+      const template = Template.fromStack(lambda.stack);
+      // Enabling log retention adds extra lambda.
+      template.resourceCountIs('AWS::Lambda::Function', 2);
+      const lambdas = template.findResources('AWS::Lambda::Function');
```

Comment:
This has been tried and adopted in [conversation handler](https://github.com/aws-amplify/amplify-backend/blob/72b2fe0d763392641d1001ebccc57c98191c24e4/packages/ai-constructs/src/conversation/conversation_handler_construct.ts#L103).
However, it was easier there because it was a new thing.

Here the problem space is constrained in such a way that:
1. In order to use custom log group (us or customer) we have to pass reference to function. So log group must be created before calling NodejsFunction ctor.
2. In order to retain log group naming we need function name first. This is not available until NodejsFunction ctor is called, making it circlular dependency problem.
3. When working on conversation handler I tried to compute function name before creating log group and nodejsfunction but I was not successful at producing function name that would match implicit cdk algorithm. That computation is internalized and I didn't find a way to implement it using available apis from cdk.
4. Changing function name and/or log group name algorithm at this point would mean destructive changes.

I wanted to avoid destructive changes, so I pursued a solution based on this extra lambda.

I think the proper solution to expose log group knobs would be to add another alternative to `LoggingOptions` in the future where we `{..existing...} | { logGroupKnobs }` .

wdyt ?

Reply:
Sounds good, at least it's only one extra function for all customer defined lambda functions.